llvm/include/llvm/BinaryFormat/Dwarf.h

```diff
@@ -613,6 +613,24 @@ enum AcceleratorTable {
   DW_hash_function_djb = 0u
 };
 
+// Uniquify the string hashes and calculate the bucket count for the
+// DWARF v5 Accelerator Table.
+inline uint32_t computeDebugNamesUniqueHashes(MutableArrayRef<uint32_t> hashes,
+                                              uint32_t &uniqueHashCount) {
+  uint32_t BucketCount = 0;
+
+  sort(hashes);
+  uniqueHashCount = llvm::unique(hashes) - hashes.begin();
```

Comment:
Without returning the iterator from `llvm::unique` or otherwise communicating info about `hashes` out to the caller - I guess the caller isn't expected to use their copy of the data structure again? Might be worth detailing in the doc comment that this function effectively consumes the `hashes` array? Or is it somehow usable afterwards?

Reply:
Yes, I want the changes to hashes to make it back to the caller.

I thought making it a MutableArray meant that the changes would get back to the user?  Originally I had it as a SmallVector &, but felipepivezan asked me to change it to a MutableArray instead.  LLVM would not allow me to declare the MutableArray as a reference?

Reply:
> LLVM would not allow me to declare the MutableArray as a reference

MutableArrays are meant to be passed by copy, they are just a "non const ptr" + "size" abstraction.

> I thought making it a MutableArray meant that the changes would get back to the user? 

The changes make it back to the user, but this is not the important distinction between MutableArrayRef and SmallVector. The SmallVector API exposes resizing capabilities, which we are not using inside this function. So you communicate this intent by using the more restricted API.

That said, David's point is that the `erase` functions only moves elements around, it doesn't actually erase them (contrary to what the name suggests). Because of that, the caller of this function can't possibly have a useful vector after this function returns.

Reply:
Ok, I went back and double check the callers; they do NOT expect a usable vector after this call, so that is fine.  However I'm getting a compile time error using the MutableArrayRef, which I could use to help figuring out how to fix: 

third_party/llvm/llvm-project/llvm/lib/CodeGen/AsmPrinter/AccelTable.cpp:43:50: error: non-const lvalue reference to type 'MutableArrayRef<uint32_t>' (aka 'MutableArrayRef<unsigned int>') cannot bind to a value of unrelated type 'SmallVector<uint32_t, 0>' (aka 'SmallVector<unsigned int, 0>')
   43 |       llvm::dwarf::computeDebugNamesUniqueHashes(Uniques, UniqueHashCount);
      |                                                  ^~~~~~~
third_party/llvm/llvm-project/llvm/include/llvm/BinaryFormat/Dwarf.h:618:74: note: passing argument to parameter 'hashes' here
  618 | inline uint32_t computeDebugNamesUniqueHashes(MutableArrayRef<uint32_t> &hashes,
      |                                                                          ^
1 error generated.

('Uniques', at the call site, line 43, is a SmallVector...)

Reply:
```
618 | inline uint32_t computeDebugNamesUniqueHashes(MutableArrayRef<uint32_t> &hashes,
```

Did you make it a reference? It should not be a reference

Reply:
Should the output parameter `uniqueHashCount` be changed to the return value?

Reply:
No, this function calculates and returns two different values:  the bucket count (for which the function is named) and the unique hash count.

Reply:
In hindsight, we should have just returned a pair

Reply:
I'll try to create another PR (probably in a day or two) to fix this. Do you know of any examples I could look at to see how this is done well?

Reply:
> With C++17 structured bindings, returning multiple values via pairs/tuples should really be the norm, rather than using reference parameters as outputs, in my opinion.

FWIW I 100% agree with you, but every now and then this runs into the LLVM's community general hesitancy to use `auto` more aggressively.

> I'll try to create another PR (probably in a day or two) to fix this. Do you know of any examples I could look at to see how this is done well?

Can't think of any example this early, but you could probably do something like:

```
inline std::pair<uint32_t, uint32_t> getDebugNamesBucketAndHashCount((MutableArrayRef<uint32_t> hashes) {
   ...
   return {BucketCount, HashCount};
}
```

And then on the call site you'd normally do something like:

```
auto [BuckerCount, HashCount] = getDebugNamesBucketAndHashCount(...)
```

However, since those are actually member variables, you can't really use a declaration like that. You'll need to store the pair in a temporary variable and assign to the member variables using `.first` and `.second`



```


```